test/test_mcmc.py

```diff
@@ -100,9 +100,12 @@ def model(data):
     data = true_coef + random.normal(random.PRNGKey(0), (1000,))
     kernel = NUTS(model=model)
     mcmc = MCMC(kernel, num_warmup=num_warmup, num_samples=num_samples)
-    mcmc.run(random.PRNGKey(2), data, collect_warmup=True)
+    mcmc.warmup(random.PRNGKey(2), data, collect_warmup=True)
```

Comment:
Hmm..this is a breaking change, and will deviate from the pyro-api interface. Can we instead do the following:
 - set `reuse_warmup_state` to `True`.
 - call `self.warmup()` from `run` only if `reuse_warmup_state` is `False` or `reuse_warmup_state = True` and `_warmup_state` is None.

We may need to move the existing logic from `.run` to another method like `.collect` or something so that `.warmup` can call that instead of `.run`.

Reply:
The problem is we don't want to have this separation for multi-chains, i.e. a chain doesn't need to wait for other chains to complete the warmup phase.

How about:
+ Making warmup a hidden method, which is only useful for benchmarking if we make the following change:
+ Return low_idx val and last val for each fori_collect run. And warmup state will be stored if `reuse_warmup_state` is False.

WDYT?

Reply:
> The problem is we don't want to have this separation for multi-chains, i.e. a chain doesn't need to wait for other chains to complete the warmup phase.

Could you explain how this is handled in the current code? 

> Return low_idx val and last val for each fori_collect run. And warmup state will be stored if reuse_warmup_state is False.

That sounds reasonable. I'm not sure if we need to return `last_val`for benchmarking, unless it is to support things like being able to restart from the last state.

More generally, I think amongst the space of solutions, we can probably rate them along the following lines:
 - Whether most of the state setting required for handling multiple chains is confined to a single method. This used to be `.run` but increasingly, things have become complex enough that the logic has started leaking out. Still, I think it will help to contain as much of the complexity to a single method so that we are able to freely refactor other parts without keeping the multi chain constraints in mind all the time.
 - Whether it is compatible with pyro / pyro-api's current interface.
 - Minimizing model re-compilation whenever possible.
 - Any benchmarking specific changes shouldn't impact the interface unless they also support a concrete use case.

Reply:
Regarding the user interface, there are two changes for this release:
+ Removing `collect_warmup` arg, which is used to diagnose the warmup phase. Most users won't use it and it can be achieved by using `_warmup` method.
+ Having an additional `reuse_warmup_state` arg, which is useful for #438 (though I prefer to have a hidden method doing this job at least for the next release, I leave it up to you to make the decision).

New hidden methods `_compiling` and `_warmup` are mainly used to separate out compiling/warmup/sampling time for benchmarking. They only affect the `fori_collect` interface.

> Could you explain how this is handled in the current code?

Currently, for each chain, we run both warmup + sampling, rather than for each chain, we run warmup and after finishing, we run sampling. The changes in this PR do not affect that. However, if users run `_warmup` first and run sampling with `reuse_warmup_phase`, then they have to wait for `_warmup` to finish.

> amongst the space of solutions

I guess the changes in this PR only affect `fori_collect` interface. So it is good to go.

> we need to return last_val for benchmarking

So far, it is not useful for:
+ compiling (because val_lower is returned and val_lower=val_last)
+ warmup with collect=False (because val_last=val_lower)
+ run with/without reuse_warmup_state

It is only useful for:
+ warmup with collect_warmup=True (where val_last=warmup_state)
+ restart run from last state (useful when users want to draw a lot of samples but do not have enough RAM, the memory restricts `collection` to have a limited size)

On the other hand, val_lower is useful for:
+ run with reuse_warmup_state

Though they affect the `fori_collect` interface, I think it is good to have both of them. Otherwise, we can make fori_collect with args to get those values a hidden function and only expose the one without `collect_size`, `return_last_val`, `return_init_val`,... args. WDYT? That way, I believe no one will blame us for having those additional args. ;) (anyway, this might not be an important point).

Reply:
So it seems like this is already doing what I mentioned, i.e. `run` will run both warmup and sampling. The only difference is that if we are looking to use the last `warmup` state, then we need to call `.warmup` explicitly. I think that's a fine constraint to have. So please disregard my comments above. I had just one question that I asked above.

Reply:
Regarding removing `reuse_warmup_state`, I think we can probably do that now? The only reason why we needed it earlier was to be able to reuse/rerun warmup, but we can do it now by just calling `.warmup` directly.

Reply:
> So it seems like this is already doing what I mentioned, i.e. run will run both warmup and sampling. The only difference is that if we are looking to use the last warmup state, then we need to call .warmup explicitly.

Yep!

> Regarding removing reuse_warmup_state, I think we can probably do that now?

I think it is fine to remove. Though if users want to do `warmup + run` again, then they have to call `.warmup` again or setting `._warmup_state = None`. WDYT?

Reply:
I think that's fine, i.e. users can continue to use `.run` which will have the same behavior as earlier. Otherwise, if they explicitly call `.warmup` first, we will skip warmup when they use `.run`.